sdk/lib/_http/websocket_impl.dart

```diff
@@ -995,12 +995,16 @@ class _WebSocketImpl extends Stream with _ServiceObject implements WebSocket {
   Timer? _closeTimer;
   _WebSocketPerMessageDeflate? _deflate;
 
-  static final HttpClient _httpClient = new HttpClient();
+  static HttpClient _httpClient = new HttpClient();
 
   static Future<WebSocket> connect(
       String url, Iterable<String>? protocols, Map<String, dynamic>? headers,
-      {CompressionOptions compression =
-          CompressionOptions.compressionDefault}) {
+      {CompressionOptions compression = CompressionOptions.compressionDefault,
+      HttpClient customClient}) {
+    //overriding _httpClient if user specifies a custom http client.
+    if (customClient != null) {
+      _httpClient = customClient..userAgent ??= _httpClient.userAgent;
```

Comment:
I think overriding static `_httpClient` here will be confusing because it means that if you call `connect` twice with different customClients the first customClient will be overridden with second one. Currently it will result in `userAgent` for the fist connect reporting the value from customClient of the last connect.

Reply:
I agree, however, with how its set up, it would require the httpClient to be tied to the webSocket itself (which I think it should be).  However that problem exists in the current implementation does it not?  So it doesn't solve that problem, but it does solve having a custom httpClient for all webSocket connections.

Reply:
> I agree, however, with how its set up, it would require the httpClient to be tied to the webSocket connection itself (which I think it should be). 

Yeah, I think having httpClient as an instance field of _WebSocketImpl would be cleaner. It could default to static final HttpClient that can be used if user didn't provide customClient.

> However that problem exists in the current implementation does it not?

Well, currently there is only one httpClient used by all web socket connections, so it being static is okay.

Reply:
In that case, would it be best to create a separate ticket for instancing the httpClient with the webSocket?  That said, I think that it would be much more clear to the user to have getters and setters for the static httpClient instead of having it passed as a method parameter.  If you guys agree, i'll update this PR to reflect that.

Reply:
If it wasn't clear by my proposal, having two separate tasks, one having the existing framework unchanged, but allowing the user to change the httpClient for all webSocket connections, and the other instancing the httpClient on a per webSocket connection basis.  Since the first is literally making a private variable accessible, it shouldn't have much, if any, technical debt on the second task.

Reply:
Main issue with exposing global writeable httpClient that is used for all websocket clients is that it will be breaking change(hard) to get rid of it.
Now I think that httpClient actually should not be an instance property of the websocket as it's only used to establish connection, only used during connect static method runtime. Also websocket can be created without httpClient at all: from regular upgraded socket.

The only question is about `userAgent` static property, which configures static(global) httpClient that is used when no customClient is specified in `connect` call.
I think it should be fine if this property only reads/updates only that default static global httpClient, doesn't affect customHttpClient user can specify in `connect`. This way the change doesn't affect existing code that updates `userAgent` property of static global httpClient and is hopefully intuitive that it won't affect custom httpClient that was provided in `connect` call.

Reply:
I still don't get how this change helps.
The class _WebSocketImpl is a private class and would be visible to code only within dart:io library, how will the
user code invoke the connect method with this new parameter.
Can we add a test case that demonstrates how this change will be effective.

Reply:
> The class _WebSocketImpl is a private class and would be visible to code only within dart:io library, how will the user code invoke the connect method with this new parameter.

Yeah, we definitely have to align `WebSocket` `connect` with `_WebSocketImpl`'s one.
With that one can fix existing test that has corresponding use case commented out:
```
diff --git a/tests/standalone/io/web_socket_error_test.dart b/tests/standalone/io/web_socket_error_test.dart
index a7fdd0ef7ca..75a8ea0c24b 100644
--- a/tests/standalone/io/web_socket_error_test.dart
+++ b/tests/standalone/io/web_socket_error_test.dart
@@ -47,9 +47,10 @@ class SecurityConfiguration {
       ? HttpServer.bindSecure(HOST_NAME, 0, serverContext, backlog: backlog)
       : HttpServer.bind(HOST_NAME, 0, backlog: backlog);
 
-  Future<WebSocket> createClient(int port) =>
-      // TODO(whesse): Add a client context argument to WebSocket.connect.
-      WebSocket.connect('${secure ? "wss" : "ws"}://$HOST_NAME:$port/');
+  Future<WebSocket> createClient(int port) {
+    return WebSocket.connect('${secure ? "wss" : "ws"}://$HOST_NAME:$port/',
+        customHttpClient: secure ? HttpClient(context: clientContext) : null);
+  }
 
   void testForceCloseServerEnd(int totalConnections) {
     createServer().then((server) {
@@ -94,7 +95,6 @@ class SecurityConfiguration {
 main() {
   asyncStart();
   new SecurityConfiguration(secure: false).runTests();
-  // TODO(whesse): WebSocket.connect needs an optional context: parameter
-  // new SecurityConfiguration(secure: true).runTests();
+  new SecurityConfiguration(secure: true).runTests();
   asyncEnd();
 }
```

Reply:
@rnewquist do you still have time to make the changes(use `customHttpClient` if it is specified, otherwise use static `httpClient_`, don't override `httpClient_` with `customHttpClient` so that static `userAgent` getter/setter only affect static `httpClient_`, add a test)?

Reply:
Yeah, I can this weekend.

Reply:
thank you, @rnewquist. Have you looked at patching tests/standalone/io/web_socket_error_test.dart along the lines of the diff in my comment above?

Reply:
you want to keep tests in standalone/io and standalone_2/io in sync so if you updated tests/standalone/io/web_socket_test.dart, we should also update tests/standalone_2/io/web_socket_test.dart.

Reply:
also take a look at try-run failures on https://dart-review.googlesource.com/c/sdk/+/200262. It is automatically synchronized with this PR, hopefully once you fix those you can request another dry run with `CQ DRY RUN` button at top right corner of that webpage.

Reply:
I fixed the standalone 2 issue, however I'm a little bit confused what you would like me to do on the web_socket_error_test.

Reply:
@aam never mind, after re-reading the code, it was really clear what you wanted me to do.  it has since been fixed

Reply:
https://github.com/dart-lang/sdk/blob/main/tests/standalone/io/web_socket_error_test.dart#L97 has a TODO to enable that test so it works with secure connection. The quote in https://github.com/dart-lang/sdk/pull/46040#discussion_r682156221 is a change that can be applied to that test(lines prefixed with '-' are deleted, with '+' - added) so that is accomplished.

Reply:
Awesome, thanks

Reply:
I do not see the option for CQ DRY RUN in the googlesource webpage unfortunately.

Reply:
@athomas do you know if this is by design where PR authors are not able to trigger dry runs in Gerrit?

Reply:
@rnewquist, I triggered the dry run.

Can you please update the commit message so it roughly follows the outline we try to adhere to for vm changes:

```[io/http] <One-liner that summarizes the change>

<Paragraph/few lines that describe what was changed, what was accomplished>

<Line that starts with TEST=... that lists tests that were used to verify correct behavior>

<Line that starts with Bug:... that references existing dart issue in github>```

Reply:
On github, gerrit, or both?

Reply:
@aam I am not authorized to change the message in gerrit, so I updated the github message.

Reply:
Thank you! Looking at how this ends up being copied into Gerrit by Copybara, I think `[io/http] <One-liner...` should be put into the subject of this PR(replacing `exposed http client for web sockets` subject) leaving only `<Paragarph... etc` in the description. 
Also word `TEST` in `Test=...` lines should be all capital and there should be no space around equal sign.

You can see how we attempt to follow this convention in https://github.com/dart-lang/sdk/commits/main

Reply:
@aam fixed, also added secure tests to web_socket_test.dart in both standalones.

Reply:
FYI, if you start with a PR it's best to keep modifying the PR and just treat Gerrit as read-only (if you modify the review, that will likely break the automatic syncing from GitHub because of the conflicting edits).

Reply:
@athomas I did not know that, sorry about that.  I don't think I have any permission to edit the gerrit, so I "think" it should good, however I will keep that in mind for any future changes that I do.